src/webgpu/shader/execution/expression/call/builtin/textureDimensions.spec.ts

```diff
@@ -0,0 +1,514 @@
+        case 'sint':
+          return `${base}_${dimensions}<i32>`;
+        case 'unfilterable-float':
+          throw new Error(`'${t.params.format}' does not support sampling`);
+      }
+    }
+
+    run(t, textureView, wgslSampledTextureType(), t.params.textureDimensionsLevel, values);
+  });
+
+g.test('depth')
+  .specURL('https://www.w3.org/TR/WGSL/#texturedimensions')
+  .desc(
+    `
+fn textureDimensions(t: texture_depth_2d) -> vec2<u32>
+fn textureDimensions(t: texture_depth_2d, level: u32) -> vec2<u32>
+fn textureDimensions(t: texture_depth_2d_array) -> vec2<u32>
+fn textureDimensions(t: texture_depth_2d_array, level: u32) -> vec2<u32>
+fn textureDimensions(t: texture_depth_cube) -> vec2<u32>
+fn textureDimensions(t: texture_depth_cube, level: u32) -> vec2<u32>
+fn textureDimensions(t: texture_depth_cube_array) -> vec2<u32>
+fn textureDimensions(t: texture_depth_cube_array, level: u32) -> vec2<u32>
+fn textureDimensions(t: texture_depth_multisampled_2d)-> vec2<u32>
+
+Parameters:
+ * t: the depth or multisampled texture
+ * level:
+   - The mip level, with level 0 containing a full size version of the texture.
+   - If omitted, the dimensions of level 0 are returned.
+`
+  )
+  .params(u =>
+    u
+      .combine('format', kAllTextureFormats)
+      .filter(u => kTextureFormatInfo[u.format].sampleType === 'depth')
+      .expand('aspect', u => aspectsForFormat(u.format))
+      .unless(u => u.aspect === 'stencil-only')
+      .expand('samples', u => samplesForFormat(u.format))
+      .beginSubcases()
+      .expand('dimensions', viewDimensions)
+      .expand('textureMipCount', textureMipCount)
+      .expand('baseMipLevel', baseMipLevel)
+      .expand('textureDimensionsLevel', textureDimensionsLevel)
+  )
+  .beforeAllSubcases(t => {
+    const info = kTextureFormatInfo[t.params.format];
+    t.skipIfTextureFormatNotSupported(t.params.format);
+    t.selectDeviceOrSkipTestCase(info.feature);
+  })
+  .fn(t => {
+    const values = testValues(t.params);
+    const texture = t.device.createTexture({
+      size: values.size,
+      dimension: textureDimensionsForViewDimensions(t.params.dimensions),
+      usage:
+        t.params.samples === 1
+          ? GPUTextureUsage.TEXTURE_BINDING
+          : GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.RENDER_ATTACHMENT,
+      format: t.params.format,
+      sampleCount: t.params.samples,
+      mipLevelCount: t.params.textureMipCount,
+    });
+    const textureView = texture.createView({
+      dimension: t.params.dimensions,
+      aspect: t.params.aspect,
+      baseMipLevel: t.params.baseMipLevel,
+    });
+
+    function wgslDepthTextureType(): string {
+      const base = t.params.samples !== 1 ? 'texture_depth_multisampled' : 'texture_depth';
+      const dimensions = t.params.dimensions.replace('-', '_');
+      return `${base}_${dimensions}`;
+    }
+
+    run(t, textureView, wgslDepthTextureType(), t.params.textureDimensionsLevel, values);
+  });
+
+g.test('storage')
```

Comment:
I'm seeing `EXCEPTION: Unexpected validation error occurred: Error while parsing WGSL: :2:37 error: read-only storage textures require the readonly_and_readwrite_storage_textures language feature, which is not allowed in the current environment`
on these tests. They should probably skip themselves when that feature isn't available.

Reply:
Fixed in https://github.com/gpuweb/cts/pull/3413.